### ISSUE TITLE
Escape characters are handled incorrectly in inline config comments

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#9366]
**Tell us about your environment**

* **ESLint Version:** v4.7.2
* **Node Version:** v8.5.0
* **npm Version:** Yarn v1.0.2

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yml
rules:
    id-match: [2, "^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$", {
        properties: true
    }]
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint id-match: [2, "^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$", {properties: true}] */

function is$2dvoid(value) {
    return value == null
}
```

**What did you expect to happen?** It to pass (as it does without the comment)

**What actually happened? Please include the actual, raw output from ESLint.**

```
/path/to/file.js
  3:10  error  Identifier 'is$2dvoid' does not match the pattern '^(([^$\\W]|\\$[a-f\\d]{2})+|[$_]\\w*|[^\\W\\d]\\w*|[A-Z]([A-Z_]*[A-Z])?)$'  id-match

✖ 1 problem (1 error, 0 warnings)
```

**Comment by not-an-aardvark** [comment]
I think this is because backslashes are being handled differently in config comments. I noticed this in https://github.com/eslint/eslint/pull/9309 but didn't end up having investigate.

**Comment by isiahmeadows** [comment]
If this helps, [here's another repro, in the web interface](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IGlkLW1hdGNoOiBbMiwgXCJeKFteJFxcXFxXXXxcXFxcJFthLWZcXFxcZF17Mn0pKyRcIl0gKi9cblxuZnVuY3Rpb24gaXMkMmR2b2lkKHZhbHVlKSB7XG4gICAgcmV0dXJuIHZhbHVlID09IG51bGxcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=).

**Comment by isiahmeadows** [comment]
Oh, and [this actually works as intended](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IGlkLW1hdGNoOiBbMiwgXCJeKFteJFxcV118XFwkW2EtZlxcZF17Mn0pKyRcIl0gKi9cblxuZnVuY3Rpb24gaXMkMmR2b2lkKHZhbHVlKSB7XG4gICAgcmV0dXJuIHZhbHVlID09IG51bGxcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=).

(Note: those two repros use a simpler regexp, but it works to the same effect.)

**Comment by Aladdin-ADD** [comment]
hmm..interesting! seems levn will replace `\\` to `\\\\`.  (we are using it to parse comment config)
https://github.com/gkz/levn/blob/a92b9acf928282ba81134b4ae8e6a5f29e1f5e1e/src/parse-string.ls#L92

**Comment by platinumazure** [comment]
I was the one who suggested using levn for inline comment handling, reasoning that it would be nice to have the same syntax in inline comments as on the command line (for options like `--rule`). I didn't realize that it would be incompatible with JavaScript/JSON string literals in some cases.

@eslint/eslint-team Should we consider removing levn from inline comment handling, given this issue? Or can we just update the inline comment docs to point to levn? Maybe we could reasonably fix this in levn?

**Comment by not-an-aardvark** [comment]
> Should we consider removing levn from inline comment handling, given this issue?

To clarify, do you mean that we would keep levn's semantics (aside from this case) without using the library? Or do you mean that we would require inline comments to actually be JSON?

I would be fine with the former (provided it isn't too much work), but I don't think we should do the latter -- it would be a big breaking change, and I like not needing to quote everything in inline comments.

> Or can we just update the inline comment docs to point to levn?

This doesn't seem ideal to me -- I think it's unexpected that a comment which appears to be JSON would not get parsed as JSON.

> Maybe we could reasonably fix this in levn?

:+1:, although it [seems unmaintained](https://github.com/gkz/levn).

**Comment by Aladdin-ADD** [comment]
https://github.com/eslint/eslint/blob/ee99876ccc76585b0af2c8673d7eb25bb0b972fa/tests/lib/linter.js#L2421

so this test should be failing -- it passed just because of the issue?

**Comment by isiahmeadows** [comment]
Yes, that is the correct understanding.

**Comment by platinumazure** [comment]
I found another example of this while trying to implement #10197. In the max-len docs, we specify a rule config inline and it crashes ESLint when trying to initialize the rule:

```js
/*eslint max-len: ["error", { "ignorePattern": "^\\s*var\\s.+=\\s*require\\s*\\(" }]*/
```

I'll try to write a report in levn in the next few days.

**Comment by platinumazure** [comment]
I've opened gkz/levn#2 to try to get some discussion going. (That said, there hasn't been any activity on levn in nearly three years, so I might need to take it a step further and write a PR.)

**Comment by kaicataldo** [comment]
How do we want to proceed with this? It unfortunately doesn't look like there has been any movement on the `levn` issue. It seems like moving away from `levn` might be the prudent thing to do.

**Comment by kaicataldo** [comment]
I think the only big difference is that string values _must_ be quoted (though it can be single or double quotes).

**Comment by not-an-aardvark** [comment]
It seems like that might be a significant breaking change, since it would make `/* eslint foo-rule: error */` fail to parse and require `/* eslint foo-rule: "error" */` instead.

Would it work to parse the strings as YAML instead? I know YAML is a superset of JSON and allows unquoted strings, although I'm not sure how it handles some edge cases.

**Comment by kaicataldo** [comment]
That's correct, parsing as JSON5 would be a breaking change. I made a [quick PoC](https://github.com/eslint/eslint/commit/56ce152549f26d9f153a45bf4be7cb29a5412dba) to help evaluate this.

> Would it work to parse the strings as YAML instead? I know YAML is a superset of JSON and allows unquoted strings, although I'm not sure how it handles some edge cases.

That's an interesting thought! Seems worth exploring.

**Comment by platinumazure** [comment]
The levn project has just issued version 0.4.1 which allows quoted string literals to use the same escaping rules as JavaScript itself, so upgrading to latest levn may resolve this issue.
